_SPIKE_.md

```diff
@@ -0,0 +1,543 @@
+
+  ```js
+  {
+    email: {
+      type: 'string',
+      unique: true,
+    }
+  }
+  ```
+
+We should also allow models to configure indexes for certain databases only by
+using database-specific options we already use to control column name, storage
+data type, etc.
+
+```js
+{
+  email: {
+    type: 'string',
+    mysql: {
+      index: true,
+      // unique: true
+    }
+  }
+}
+```
+
+### Indexes at model level
+
+Problem 1: different databases provide different indexing features.
+
+- It seems that all databases support ASC/DESC option at column-level.
+- MongoDB adds extra column-level option `text` as an alternative to ASC/DESC.
+- PostgreSQL allows column expressions, e.g. `lower(email)`.
+
+Problem 2: do we want to use database column names or LB property names? These
+two names can be different, especially when using database-specific config to
+map property names to custom column names.
+
+**The proposal:**
+
+At high-level, keep the current syntax where indexes are defined via a key-value
+map stored in `settings.indexes` property, the key is the index name and the
+value is an index definition object.
+
+```js
+{
+  strict: false,
+  forceID: true,
+  indexes: {
+    uniqueEmail: {
+      // index definition
+    },
+    nameQueries: {
+      // index definition
+    }
+  }
+}
+```
+
+Individual indexes can be defined as follows:
+
+- Add a new field `properties` as a key-value map from property names to
+  indexing order:
+
+  ```js
+  // definition of an individual index
+  {
+    properties: {
+      email: 1, // ASC
+      createdAt: 'DESC', // alias for -1
+      bio: 'text', // database-specific value (MongoDB's "text")
+    }
+  }
+  ```
+
+  Important: property names are mapped to database column names when building
+  the index definition.
+
+- Keep supporting `keys` field as a key-value map from database column name to
+  indexing order, see the description of the actual status below. Entries from
+  `keys` should me merged with entries from `properties`, `keys` taking
+  precedence (replacing `properties` entries).
+
+- Keep supporting `unique` field (set it to `true` to let the index enforce
+  uniqueness).
+
+- Database-specific options will be stored under a key with the connector name:
+
+  ```js
+  {
+    properties: {
+      email: 'ASC',
+    },
+    mongodb: {
+      sparse: true,
+    },
+    mysql: {
+      kind: 'fulltext',
+      type: 'hash',
+    },
+    postgresql: {
+      type: 'hash',
+    }
+  }
+  ```
+
+- To support more complex indexes offered by e.g. PostgreSQL, we should support
+  `columns` field too. This field should contain an array of column-index
+  specification, for example:
+
+  ```js
+  ['email ASC', 'lower(name) DESC'];
+  ```
+
+  In the contrary with the current implementation, I am proposing that `columns`
+  should replace any column list created from `properties` and `keys`, and that
+  `columns` should be nested inside connector-specific options:
```

Comment:
I agree with the syntax for `postgresql.columns = ['lower (email) ASC`]`. Just want to make sure what will happen if a user defines both, `properties` and `postgreqsl.columns`. 

It seems like the connector will just use whatever is defined in the connector property. I don't know how you could know programmatically that `[lower(email) ASC` was intended to replace `email`

Reply:
>  I don't know how you could know programmatically that `lower(email) ASC` was intended to replace` email`.

Exactly! That's why I am proposing that `columns` should take precedence over `keys`/`properties` and replace anything defined via those key/value maps.

- If `postgresql.columns` is defined, then the connector uses `columns` and ignored `keys` & `properties`.
- If `postgresql.columns` is not defined, the connector will merge entries defined by `keys` and `properties`.

Reply:
> If postgresql.columns is not defined, the connector will merge entries defined by keys and properties.

So the order is `postgresql.columns` -> `keys` -> `properties`, correct?

Reply:
Yes.
If `columns` are set, then `keys` and `properties` are ignored completely.

Otherwise the content from `keys` and `properties` is merged, e.g. this will create a composite index over two columns.

```js
{
  keys: {
    name: 'ASC',
  },
  properties: {
    email: 'DESC',
  }
}
```